examples/md-flexible/src/ParallelVtkWriter.cpp

```diff
@@ -139,9 +139,47 @@ void ParallelVtkWriter::recordParticleStates(size_t currentIteration,
 
   // print positions
   timestepFile << "        <DataArray Name=\"positions\" NumberOfComponents=\"3\" format=\"ascii\" type=\"Float32\">\n";
+  const auto boxMax = autoPasContainer.getBoxMax();
   for (auto particle = autoPasContainer.begin(autopas::IteratorBehavior::owned); particle.isValid(); ++particle) {
-    auto pos = particle->getR();
-    timestepFile << "        " << pos[0] << " " << pos[1] << " " << pos[2] << "\n";
+    // When we write to the file in ASCII, values are rounded to the precision of the filestream.
+    // Since a higher precision results in larger files because more characters are written,
+    // and mdflex is not intended as a perfectly precice tool for application scientists,
+    // we are fine with the rather low default precision.
+    // However, if a particle is very close to the domain border it can happen that the particle position is rounded
+    // exactly to the boundary position. This then causes problems when the checkpoint is loaded because boxMax is
+    // considered to be not part of the domain, hence such a particle would not be loaded and thus be lost.
+    // This function identifies these problematic values and raises the write precision just for this value high enough
+    // to be distinguishable from the boundary.
+    const auto writeWithDynamicPrecision = [&](double d, double border) {
+      const auto initialPrecision = timestepFile.precision();
+      // Simple and cheap check if we even need to do anything.
+      if (border - d < 0.1) {
+        using autopas::utils::Math::roundFloating;
+        // As long as the used precision results in the two values being the same increase it
+        while (roundFloating(d, timestepFile.precision()) == border) {
```

Comment:
As far as I understand it, this can only become a problem if the border is a value that is an infinite sequence that can not be represented exactly by double like e.g. `1/3`. Finite values that can not be represented like `0.3` are fine.

~I'm not sure we can somehow come into this position for the total simulation box.~ Actually md-flex and ls1 can't (the parsers don't support this) but theoretically some other other program could set this.
For inner mpi-boundaries this doesn't matter because particles then are simply loaded into the neighbor rank.

I'll fix it.

Reply:
-> 3a165ca0d

Reply:
Here my code to reproduce my tests to check that this makes sense:

<details><summary><code>leCode.cpp</code></summary>
<p>

```cpp
#include <cmath>
#include <iomanip>
#include <iostream>
#include <vector>

double roundTo_math(double d, int precision) {
  const auto scale = std::pow(10, precision);
  return std::round(d * scale) / scale;
}

double roundTo_mathFloat(double d, int precision) {
  if (d == 0.0) return d;  // Special case for 0 to avoid log10 error
  const auto factor = std::pow(10, precision - std::ceil(std::log10(std::abs(d))));
  return std::round(d * factor) / factor;
}

double roundTo_ss(double d, int precision) {
  std::stringstream ss;
  ss << std::fixed << std::setprecision(precision);
  ss << d;
  double ret;
  ss >> ret;
  return ret;
}

double roundTo_ssstod(double d, int precision) {
  std::stringstream ss;
  ss << std::fixed << std::setprecision(precision);
  ss << d;
  return std::stod(ss.str());
}

double roundTo_sprintf(double d, int precision) {
  char buf[7];
  snprintf(buf, 7, "%.2f", d);
  return std::stod(buf);
}

bool isNear(double a, double b, double relativeDifference) {
  const auto greaterNumber = std::max(std::abs(a), std::abs(b));
  const auto absoluteDifference = relativeDifference * greaterNumber;
  const auto diff = std::abs(a - b);
  return diff <= absoluteDifference;
}

template<class X = int>
using vec = std::vector<X>;

int main() {
  const double border = 1/3.;
  const double d = border - 0.000000001;
  std::cout << "Naive"
            << "\n";
  std::cout << "Border            : " << border << "\n";
  std::cout << "operator<<        : " << d << "\n";
  std::cout << std::setprecision(20) << std::boolalpha;
  std::cout << "Precision 20"
            << "\n";
  std::cout << "Border            : " << border << "\n";
  std::cout << "operator<<        : " << d << "\n";

  for (int precision = 6; precision < 25; ++precision) {
    std::cout << std::setprecision(precision);
    std::cout << "Precision : " << std::setw(2) << std::right << precision
              << " | roundTo_mathFloat : " << std::setw(26) << std::left << roundTo_mathFloat(d, precision)
              << " | is equal to border: " << std::setw(5) << std::right<< (roundTo_mathFloat(d, precision) == border)
              << " | is near border: " << std::setw(5) << std::right<< (std::abs(roundTo_mathFloat(d, precision) - border) < std::pow(10, -precision))
              << " | diff: " << std::setw(30) << std::right<< std::abs(roundTo_mathFloat(d, precision) - border)
              << " | max tolerance: " << std::setw(30) << std::right<< std::pow(10, -precision - 1) << "\n";
  }
}

```

</p>
</details>